tensorboard/components/tf_backend/test/requestManagerTests.ts

```diff
@@ -289,6 +290,83 @@ describe('backend', () => {
         // resolving the first request triggers finishTheTest
       });
     });
+
+    it('throws InvalidRequestOptionsError on bad RequestOptions', function() {
+      const rm = new RequestManager();
+        const badOptions = new RequestOptions();
+      badOptions.methodType = HttpMethodType.GET;
+      badOptions.body = "a body";
+      chai.assert.throws(
+        ()=>rm.requestWithOptions("http://www.google.com", badOptions),
+        InvalidRequestOptionsError);
+    });
+
+    describe('tests using sinon.fakeServer', function() {
+      let server;
+      
+      beforeEach(function() {
+        server = sinon.fakeServer.create();
+        server.respondImmediately = true;
+        server.respondWith("{}");
+      });
+
+      afterEach(function() {
+        server.restore();
+      });
+      
+      it('builds correct XMLHttpRequest when request(url) is called',
+         function() {
+           const rm = new RequestManager();
+           return (rm.request("my_url") as any)
+             .finally(()=>{
+               chai.assert.lengthOf(server.requests, 1);
+               chai.assert.equal(server.requests[0].url, "my_url");
+               chai.assert.equal(server.requests[0].requestBody, null);
+               chai.assert.equal(server.requests[0].method, HttpMethodType.GET);
+               chai.assert.notProperty(server.requests[0].requestHeaders,
+                                       "Content-Type");
+             }).finally(()=>{
+               server.restore();
+             });
+         });
+
+      it('builds correct XMLHttpRequest when request(url, postData) is called',
+         function() {
+           const rm = new RequestManager();
+           return (rm.request("my_url",
+                              {"key1": "value1", "key2": "value2"}) as any)
+             .finally(() => {
+               chai.assert.lengthOf(server.requests, 1);
+               chai.assert.equal(server.requests[0].url, "my_url");
+               chai.assert.equal(server.requests[0].method,
+                                 HttpMethodType.POST);
+               chai.assert.instanceOf(server.requests[0].requestBody, FormData);
+               chai.assert.sameDeepMembers(
+                 Array.from((server.requests[0].requestBody as any).entries()),
+                 [["key1", "value1"], ["key2", "value2"]]);
+             });
+         });
+
+      it('builds correct XMLHttpRequest when requestWithOptions is called',
+         function() {
+           const rm = new RequestManager();
+           const requestOptions = new RequestOptions();
+           requestOptions.methodType = HttpMethodType.POST;
+           requestOptions.contentType = "text/plain;charset=utf-8";
+           requestOptions.body = "the body";
+           return (rm.requestWithOptions("my_url", requestOptions) as any)
+             .finally(()=>{
```

Comment:
JW, does `.then` better capture what you think?

Reply:
I want the assertions to be called even if the promise rejects.

Reply:
Make sense in perspective of the server.

but that sounds like there should be a test around result (whether the promise resolves or not) and add a case when request fails to resolve too.

Reply:
Actually, thinking some more about this, I agree that 'then' is better'. Changed.